### DISCUSSION TITLE
Accessing/calculating flow gradients

### DISCUSSION THREAD

**Discussion by JayStandridge** [AMReX-Combustion/PeleC#555] [Q&A]
Howdy,

I was wondering how I can access (or calculate) the flow gradients. I have tried defining new  derived variables and assigning functions to them. I have used pc_derdivu as a template as to how I might calculate the velocity gradient tensor, but I was wondering if that was calculated elsewhere in the code, and how I might be able to access it. Eventually, I will need to calculate the gradients of the mole fractions (currently running the PMF regtest, as my advisor is wanting me to work with hydrogen combustion), and the temperature gradient. Any help, advice, or ideas are appreciated.

Thank you

**Comment by drummerdoc** [comment]
Hey Jay, Probably one of the worst things about dealing with an AMR code is reproducing gradient calculations used in the code to advance the solution - the stencils at the coarse-fine boundary are hideous (and are typically implemented in a complex sequence of steps difficult to redo without copying massive sections of code.  

I'd suggest you create an "Issue" requesting access to flux variables, as used by the code.  It might take a little time to get exactly what you want, but it would be good to involve some of the project team in designing an interface for this that is workable.

**Reply by JayStandridge** [reply]
Thank you for the reply- yes, in my first attempt at calculating the gradients, the cells adjacent to a refinement zone are giving me issues.

I will submit an "Issue", thank you for the suggestion.

**Comment by marchdf** [comment]
Do you just want these gradients for post-processing/visualization or are you actually wanting to use them in the code?

**Reply by JayStandridge** [reply]
I am mostly aimed at post-processing, however, in trying to familiarize myself with the code, I believed the easiest way (I am a beginner at best with C++) to visualize the data would be to edit Derive.H/cpp and setup.cpp to define new derived variables that I can subsequently visualize with paraview. Is there a better way to achieve this?

**Comment by drummerdoc** [comment]
It occurred to me that there are some special quantities computed during the integration that users may want access to, such as the gradients.  One approach is to add some code to optionally save those off if requested and then move those into the derived variables before output.  It might lead to complicated internal logic, but it seems actually pretty useful, especially for debugging.  Wondering what the team thinks about this...
